File: lab3/cs/DiffieHellman/dh/DHUser.cs
namespace dh;

using dh.Interfaces;

public class DHUser<T> where T : IGaloisField<T> {
    public DHUser(IGaloisFieldFactory<T> tFactory, DHSetup<T> setup) {
        this.tFactory = tFactory;
        this.setup = setup;
        this.secret = Math.Max(DHUser<T>.randomULong(), 1UL);
        this.privKey = default(T);
    }

    public T GetPublicKey() {
        return this.setup.Power(this.setup.Generator, this.secret);
    }

    public void SetKey(T key) {
        this.privKey = this.setup.Power(key, this.secret);
    }

    public T Encrypt(T message) {
        if (this.privKey == null)
            throw new InvalidOperationException("Cannot encrypt a message without a private key");

        return message.Multiply(this.privKey!);
    }

    public T Decrypt(T code) {
        if (this.privKey == null)
            throw new InvalidOperationException("Cannot decrypt a code without a private key");

        return code.Divide(this.privKey!);
    }

    private static ulong randomULong() {
        byte[] buffer = new byte[8];
        DHUser<T>.random.NextBytes(buffer);
        return BitConverter.ToUInt64(buffer, 0);
    }

    private static Random random = new Random(42);

    private IGaloisFieldFactory<T> tFactory;
    private DHSetup<T> setup;
    private ulong secret;
    private T? privKey;
}

File: lab3/java/DiffieHellman/src/main/java/com/jpp_lab3/app/DHUser.java
package com.jpp_lab3.app;

import com.jpp_lab3.app.interfaces.*;

import java.util.Optional;
import java.util.Random;

public class DHUser<T extends IGaloisField<T>> {
    DHUser(IGaloisFieldFactory<T> tFactory, DHSetup<T> setup) {
        this.tFactory = tFactory;
        this.setup = setup;
        this.secret = Math.max(Math.abs(DHUser.random.nextLong()), 1);
        this.privKey = Optional.empty();
    }

    public T getPublicKey() {
        return this.setup.power(this.setup.getGenerator(), this.secret);
    }

    void setKey(T key) {
        this.privKey = Optional.of(this.setup.power(key, this.secret));
    }

    public T encrypt(T message) throws IllegalStateException {
        if (!this.privKey.isPresent())
            throw new IllegalStateException("Cannot encrypt a message without a private key");

        T encryptedMessage = this.tFactory.clone(message);
        encryptedMessage.multiply(this.privKey.get());
        return encryptedMessage;
    }

    public T decrypt(T message) throws IllegalStateException {
        if (!this.privKey.isPresent())
            throw new IllegalStateException("Cannot decrypt a code without a private key");

        T decryptedMessage = this.tFactory.clone(message);
        decryptedMessage.divide(this.privKey.get());
        return decryptedMessage;
    }

    private static Random random = new Random(42);

    private IGaloisFieldFactory<T> tFactory;
    private DHSetup<T> setup;
    private long secret;
    private Optional<T> privKey;
}
